File: gorays/result.go
package main

import (
	"encoding/json"
	"log"
	"os"
)

type results []float64

func (r results) Sum() (sum float64) {
	for _, s := range r {
		sum += s
	}

	return
}

func (r results) Average() float64 {
	return r.Sum() / float64(len(r))
}

func (r results) Save() {
	f, err := os.Create(*resultfile)
	if err != nil {
		log.Panic(err)
	}
	defer f.Close()

	data := struct {
		Average float64   `json:"average"`
		Samples []float64 `json:"samples"`
	}{
		r.Average(),
		r,
	}

	enc := json.NewEncoder(f)
	enc.Encode(data)
}

File: rustrays/result.rs
extern mod extra;

use std::io::fs::File;
use extra::json::PrettyEncoder;
use extra::serialize::Encodable;

pub struct Result {
	samples: ~[f64]
}

#[deriving(Encodable)]
struct JsonResult<'a> {
	average: f64,
	samples: &'a[f64]
}

impl Result {
	pub fn sum(&self) -> f64 {
		let mut sum = 0.0;
		for &s in self.samples.iter() {
			sum += s;
		}
		sum
	}

	pub fn average(&self) -> f64 {
		self.sum() / self.samples.len() as f64
	}

	pub fn save(&self, resultfile: &str) {
		match File::create(&Path::new(resultfile)) {
			Some(ref mut file) => {
				let mut enc = PrettyEncoder::new(file);
				JsonResult { average: self.average(), samples: self.samples }.encode(&mut enc);
				file.flush();
			}
			None => {
				println(format!("Opening {} file failed!", resultfile));
			}
		};
	}
}
